### DISCUSSION TITLE
[Feature] Support URI template specification

### DISCUSSION THREAD

**Discussion by luisaveiro** [wolfpackthatcodes/typescript-http-client#181] [Ideas]
#### Summary
Allow a developer to construct request URLs using the [URI template specification](https://www.rfc-editor.org/rfc/rfc6570).

---

#### Possible implementation
The HTTP Client provides a `withUrlParameters` method that defines the URL parameters that can be expanded by the URI template.

##### Method signature

```typescript
export default class HttpClient {
  /**
   * Specify the URL parameters that can be substituted into the request URL.
   *  
   * @param {Object.<string, boolean | number | string>} urlParameters
   * 
   * @returns {this}
   */
  withUrlParameters(urlParameters: { [key: string]: boolean | number | string }): this {
    // code ...
  }
}
```

---

#### Code examples

```typescript
import { HttpClient } from '@wolfpackthatcodes/http-client';

const response = new HttpClient()
  .withUrlParameters({
    domain: 'https://api.example.local',
    endpoint: 'test',
  })
  .get('{+domain}/{endpoint}');
```

**Comment by ageddesi** [comment]
This is a great idea and something I would like to help on. 
What is the + symbol for in your example above?

**Reply by luisaveiro** [reply]
@ageddesi according to [URI template specification: 3.2.3.  Reserved Expansion: {+var}](https://www.rfc-editor.org/rfc/rfc6570#section-3.2.3).

> Reserved expansion, as indicated by the plus ("+") operator for Level 2 and above templates, is identical to simple string expansion except that the substituted values may also contain pct-encoded triplets and characters in the reserved set.
